libvast/src/system/index.cpp

```diff
@@ -332,6 +306,34 @@ void index_state::create_active_partition() {
   active_partition.capacity = partition_capacity;
   active_partition.id = id;
   VAST_DEBUG("{} created new partition {}", self, id);
+  // Move responsibility for fulfilling the flush promise to the active
+  // partitions.
+  if (flush_promises)
+    for (auto&& flush_promise : std::exchange(*flush_promises, {}))
```

Comment:
I don't quite understand how this works - it looks like all `flush_promises` are cleared out every time we create an active partition? That would not work in case we create several partitions in a row before shutting down.

Reply:
It simply moves them over from the `shared_ptr<vector<rp>>` from inside the index to those inside the active partitions, and then takes the last fulfilled rp from the active partitions and fulfills the original promise with it.

Reply:
Yeah, but how do the promises get back out one active partition to the next? Consider the sequence

```
1. Subscribe arrives in the index
2. An active partition P1 is created, the response promised is moved inside
3. More data arrives, the old active partition is decomissioned and another one P2 is created
4. Stream ends
```

Now P2 should deliver the response promises after its stream is clean, but how does it get them?

Reply:
Step 2 in your sequence is wrong; the `forwarded_flush_promises` are not moved into the active partition, but rather shared with it, and when a new active partition is created it also receives all not yet fulfilled `forwarded_flush_promises`.